### PR TITLE
Add CLI for benchmarking datasets on SAM models

### PR DESCRIPTION
**Pull request by anwai98**
The prototype CLI for benchmarking microscopy datasets on SAM models is ready for the first round of review (tested on LIVECell and DeepBacs)

I'll work on testing it on 3d datasets and extend interactive segmentation functionality for 3d (using SAM)

### CODE REVIEW

File: micro_sam/evaluation/benchmark_datasets.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        zip(image_paths, gt_paths), total=len(image_paths),
+        desc=f"Extracting patches for {dataset_choice}"
+    ):
+        image = imageio.imread(image_path)
+        gt = imageio.imread(gt_path)
+
+        if len(np.unique(gt)) == 1:  # There could be labels which does not have any annotated foreground.
+            continue
+
+        tiling = blocking([0, 0], gt.shape, tile_shape)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:209]: This will not work for 3d data, we need to passs `[0, 0, 0]` for this. It's probably best to define a variable `origin` for this. See above.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+            available_datasets[choice]()
+        else:
+            raise ValueError(f"'{choice}' is not a supported choice of dataset.")
+
+    return dataset_choice
+
+
+def _extract_slices_from_dataset(path, dataset_choice):
+    if dataset_choice in LM_2D_DATASETS or EM_2D_DATASETS:
+        ndim, tile_shape = 2, (512, 512)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:170]: `origin = [0, 0]`

File: micro_sam/evaluation/benchmark_datasets.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    else:
+        ndim, tile_shape = 3, (32, 512, 512)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:172]: `origin = [0, 0, 0]`

File: micro_sam/evaluation/benchmark_datasets.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+        image = imageio.imread(image_path)
+        gt = imageio.imread(gt_path)
+
+        if len(np.unique(gt)) == 1:  # There could be labels which does not have any annotated foreground.
+            continue
+
+        tiling = blocking([0, 0], gt.shape, tile_shape)
+        n_tiles = tiling.numberOfBlocks
+        tiles = [tiling.getBlock(tile_id) for tile_id in range(n_tiles)]
+        crop_boxes = [[tile.begin[1], tile.begin[0], tile.end[1], tile.end[0]] for tile in tiles]
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:212]: Again, needs some update for 3d.

Comment (constantinpape) [reply micro_sam/evaluation/benchmark_datasets.py:212]: It would be easiest if you just create a list of `tuple[slice]` here directly and then use it wherever relevant below. This would enable 3d easily and also reduce confusion with flipping around the axis orders below.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+            device=device,
+            checkpoint_path=checkpoint_path,
+            ndim=ndim,
+            verbose=False,
+            **auto_seg_kwargs
+        )
+
+    prediction_paths = natsorted(glob(os.path.join(prediction_dir, "*")))
+    res = run_evaluation(gt_paths=gt_paths, prediction_paths=prediction_paths, save_path=result_path)
+    print(res)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:300]: We should create a dataframe and then save it as csv.

**Review by constantinpape — COMMENTED** [review]
Looks good overall. Two minor points:
- We can make the data extraction work easily for 2d and 3d, see comments.
- We should define a csv output that everything is saved to for easy further analysis

File: micro_sam/evaluation/benchmark_datasets.py (hunk 5 above)
Comment (anwai98) [reply micro_sam/evaluation/benchmark_datasets.py:324]: The results are stored at `result_path` filepath, passed to the `save_path` argument. I'll refactor the evaluation for clarity. Thanks for pointing out!

**Comment by anwai98** [comment]
Okay, in principle the pipeline is ready now.

@constantinpape let me know how it looks now.

PS: There are a few major bulks I still need to take care of:
- Run both 2d and 3d segmentation for volumetric data.
- Allow user to flexibly choose between AMG / AIS.
- (Optional) Remove the cached original inputs.
- (Optional) Remove the cached preprocessed inputs.
- (Optional) Remove the stored predictions.

**Comment by constantinpape** [comment]
> * Allow user to flexibly choose between AMG / AIS.

I think the best would be to enable running both AMG and AIS, and by default only run AIS.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    for per_paths in tqdm(paths_set, total=len(paths_set), desc=f"Extracting patches for {dataset_choice}"):
+        if ndim == 2:
+            image_path, gt_path = per_paths
+            image, gt = util.load_image_data(image_path), util.load_image_data(gt_path)
+        else:
+            image_path = per_paths
+            image = util.load_image_data(image_path, DATASET_CONTAINER_KEYS[dataset_choice][0])
+            gt = util.load_image_data(image_path, DATASET_CONTAINER_KEYS[dataset_choice][1])
+
+        skip_smaller_shape = (image.shape > tile_shape)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:230]: Does this work? We need to do some elementwise comparison here, but `image.shape` and `tile_shape` are tuples. I think you need to case them to arrays and then use `.all`:
```python
skip_smaller_shape = (np.array(image.shape) >= np.array(tile_shape)).all()
```

File: micro_sam/evaluation/benchmark_datasets.py (hunk 5 above)
Comment (constantinpape) [reply micro_sam/evaluation/benchmark_datasets.py:329]: Ok. I think we should also remove the prints of results here at the end. Otherwise the output gets confusing.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+        output_folder: Filepath to the folder where we store all the results.
+        ndim: The number of input dimensions.
+        checkpoint_path: The filepath for stored checkpoints.
+        auto_seg_kwargs: Additional arguments for automatic segmentation parameters.
+    """
+    result_path = os.path.join(output_folder, "results", "automatic_segmentation.csv")
+    prediction_dir = os.path.join(output_folder, "automatic_instance_segmentation", "inference")
+    os.makedirs(prediction_dir, exist_ok=True)
+
+    for image_path in tqdm(image_paths, desc="Run automatic segmentation"):
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:310]: This is currently a bit ineffecient: we load the model again for each image.I think for running AIS that can be quite some overhead.

Comment (constantinpape) [reply micro_sam/evaluation/benchmark_datasets.py:310]: Not sure how exactly we want to address it. One option would be enabling passing multiple paths to `automatic_instance_segmentation`.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+                volume=imageio.imread(image_path),
+                ground_truth=imageio.imread(gt_path),
+                model_type=model_type,
+                checkpoint_path=checkpoint_path,
+                # save_path=os.path.join(prediction_dir, os.path.basename(image_path)),
+                device=device,
+                interactive_seg_mode=prompt_choice,
+                min_size=10,
+            )
+            print(msa)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:396]: Remove prints for finalization.

File: micro_sam/evaluation/benchmark_datasets.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+
+        prediction_paths = natsorted(glob(os.path.join(prediction_dir, "*")))
+        res = run_evaluation(
+            gt_paths=gt_paths, prediction_paths=prediction_paths,
+            save_path=os.path.join(output_folder, "results", f"interactive_segmentation_3d_with_{prompt_choice}.csv")
+        )
+        print(res)
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:403]: Remove prints for finalization.

File: micro_sam/evaluation/benchmark_datasets.py (hunk 6 above)
Comment (anwai98) [reply micro_sam/evaluation/benchmark_datasets.py:230]: Oh yeah you are right. There's the case where the first dimension for image shape might be higher while the later matching, then this would fail. Your suggestion would work for all dimensions. Thanks for the mention!

File: micro_sam/evaluation/benchmark_datasets.py (hunk 7 above)
Comment (anwai98) [reply micro_sam/evaluation/benchmark_datasets.py:310]: I think I'll update `automatic_instance_segmentation` so that it can consider multiple paths. I'll take care of it. Thanks for the suggestion!

Comment (anwai98) [reply micro_sam/evaluation/benchmark_datasets.py:322]: It was slightly bit more complicated to enable multiple paths in the existing function, so I refactored a few things to avoid doing redundant things (eg. calling the predictor and segmenter for every image). This part should be taken care of now.

**Comment by anwai98** [comment]
This PR is now ready (tested this on DeepBacs and Lucchi).

I am now working on finalizing all open-source datasets in the preprocessing part, and with that should be finally good to go from my side.

Resolves https://github.com/computational-cell-analytics/micro-sam/issues/719

File: test/test_automatic_segmentation.py [hunk 10]
```diff
@@ -65,46 +65,63 @@ def tearDown(self):
             import torch.mps
             torch.mps.empty_cache()
 
+    # get the predictor and segmenter.
+    def _get_predictor_and_segmenter(self, ais=False, is_tiled=False, amg_kwargs={}):
```
Comment (constantinpape) [inline test/test_automatic_segmentation.py:69]: I think it would make sense to refactor this to `automatic_instance_segmentation.py` as a proper documented function to make it easy for users to load the predictor and segmenter.

File: micro_sam/automatic_segmentation.py [hunk 11]
```diff
@@ -179,17 +166,28 @@ def _convert_argval(value):
...
+    predictor, state = util.get_sam_model(
+        model_type=args.model_type, device=device, checkpoint_path=args.checkpoint, return_state=True,
+    )
+
+    segmenter = get_amg(
+        predictor=predictor, is_tiled=args.tile_shape is not None,
+        decoder=get_decoder(
+            image_encoder=predictor.model.image_encoder, decoder_state=state["decoder_state"], device=device
+        ) if "decoder_state" in state and not args.amg else None
+    )
```
Comment (constantinpape) [inline micro_sam/automatic_segmentation.py:179]: Use the refactored function here, see prev. comment

File: micro_sam/evaluation/benchmark_datasets.py [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+        model_type=model_type, device=device, checkpoint_path=checkpoint_path, return_state=True,
+    )
+
+    segmenter = get_amg(
+        predictor=predictor,
+        is_tiled=False,
+        decoder=get_decoder(
+            predictor.model.image_encoder, state["decoder_state"], device
+        ) if "decoder_state" in state and not run_amg else None
+    )
```
Comment (constantinpape) [inline micro_sam/evaluation/benchmark_datasets.py:426]: Can also use the refactored function here.

**Review by constantinpape — COMMENTED** [review]
I think we should refactor the loading functionality for automatic segmentation, rest looks good from my end now.

**Comment by anwai98** [comment]
Hi @constantinpape,

This PR is GTG from my side now. (I took care of the minor TODOs which I procrastinated since last week). Thanks!

**Review by constantinpape — APPROVED** [review]
